src/chainstate/stacks/db/accounts.rs

```diff
@@ -0,0 +1,829 @@
+            let nonce = db.get_account_nonce(principal);
+            Ok(StacksAccount {
+                principal: principal.clone(),
+                stx_balance,
+                nonce
+            })
+        }).expect("FATAL: failed to query account")
+    }
+
+    pub fn get_account_ft<'a>(clarity_tx: &mut ClarityTx<'a>, contract_id: &QualifiedContractIdentifier, token_name: &str, principal: &PrincipalData) -> Result<u128, Error> {
+        clarity_tx.connection().with_clarity_db_readonly(|ref mut db| {
+            let ft_balance = db.get_ft_balance(contract_id, token_name, principal)?;
+            Ok(ft_balance)
+        })
+        .map_err(Error::ClarityError)
+    }
+
+    pub fn get_account_nft<'a>(clarity_tx: &mut ClarityTx<'a>, contract_id: &QualifiedContractIdentifier, token_name: &str, token_value: &Value) -> Result<PrincipalData, Error> {
+        clarity_tx.connection().with_clarity_db_readonly(|ref mut db| {
+            let nft_owner = db.get_nft_owner(contract_id, token_name, token_value)?;
+            Ok(nft_owner)
+        })
+        .map_err(Error::ClarityError)
+    }
+
+    /// Called each time a transaction is invoked from this principal, to e.g.
+    /// debit the STX-denominated tx fee or transfer/burn STX.
+    /// DOES NOT UPDATE THE NONCE
+    pub fn account_debit<'a>(clarity_tx: &mut ClarityTx<'a>, principal: &PrincipalData, amount: u64) {
+        clarity_tx.connection().with_clarity_db(|ref mut db| {
+            let cur_balance = db.get_account_stx_balance(principal);
+            
+            // last line of defense: if we don't have sufficient funds, panic.
+            // This should be checked by the block validation logic.
+            if cur_balance < (amount as u128) {
+                panic!("Tried to debit {} from account {} (which only has {})", amount, principal, cur_balance);
+            }
+
+            let final_balance = cur_balance - (amount as u128);
+            db.set_account_stx_balance(principal, final_balance);
+            Ok(())
+        }).expect("FATAL: failed to debit account")
+    }
+
+    /// Called each time a transaction sends STX to this principal.
+    /// No nonce update is needed, since the transfer action is not taken by the principal.
+    pub fn account_credit<'a>(clarity_tx: &mut ClarityTx<'a>, principal: &PrincipalData, amount: u64) {
+        clarity_tx.connection().with_clarity_db(|ref mut db| {
+            let cur_balance = db.get_account_stx_balance(principal);
+            let final_balance = cur_balance.checked_add(amount as u128).expect("FATAL: account balance overflow");
+            db.set_account_stx_balance(principal, amount as u128);
+            Ok(())
+        }).expect("FATAL: failed to credit account")
+    }
+    
+    /// Verify that this transaction is not a replay, and update its nonce.
+    pub fn update_account_nonce<'a>(clarity_tx: &mut ClarityTx<'a>, tx: &StacksTransaction, account: &StacksAccount) {
```

Comment:
Where is the nonce of the transaction checked?

Reply:
I don't think it is (oops).  Will fix -- it should be checked in `process_transaction()` in `src/chainstate/stacks/db/transactions.rs`.

Reply:
Fixed in ef2e7c6c